### ISSUE TITLE
Inconsistent Types in Package URL (purl) for Ruby causing no vulns to be returned in REST API

### ISSUE THREAD

**Issue by csansone-handy** [OSSIndex/vulns#19]
There are some inconsistencies in the way Ruby gems are listed as components and this is causing issues when tracing the lineage of a vulnerability back to the components affected by the vulnerability.

For example we have the following components for the Nokogiri gem (note the difference in the _type_ used for the PURL of each component: `gem` vs. `rubygem`):

- https://ossindex.sonatype.org/component/pkg:gem/nokogiri
- https://ossindex.sonatype.org/component/pkg:rubygem/nokogiri

Both component pages say they are vulnerable to CVE-2019-11068 (amongst other vulnerabilities/advisories) that affect Nokogiri.

pkg:gem/nokogiri component is listed as vulnerable to CVE-2019-11068:
![image](https://user-images.githubusercontent.com/51127567/61301899-dcaf3d00-a7b2-11e9-8d8b-1568707a01be.png)

pkg:rubygem/nokogiri component is listed as vulnerable to CVE-2019-11068:
![image](https://user-images.githubusercontent.com/51127567/61302030-1c762480-a7b3-11e9-8967-dc645f9ea2bd.png)

However, if you go to the [vulnerability specific page for CVE-2019-11068](https://ossindex.sonatype.org/vuln/1814284c-75a0-42e1-83b6-0395d74fbafb) it does not list `pkg:gem/nokogiri` as a vulnerable component and only lists `pkg:rubygem/nokogiri`.
![image](https://user-images.githubusercontent.com/51127567/61301728-7f1af080-a7b2-11e9-8cc1-3de86b03e451.png)

This causes issues when you try to use the REST API. For example when you make an HTTP GET to https://ossindex.sonatype.org/api/v3/component-report/pkg:rubygem/nokogiri@1.8.0 CVE-2019-11068 (amongst other vulnerabilities/advisories) is returned but if you make an HTTP GET to https://ossindex.sonatype.org/api/v3/component-report/pkg:gem/nokogiri@1.8.0 an empty array/list of vulnerabilities is returned. Again note the difference in the _type_ used for the PURL of each component _coordinate_ supplied to the REST API: `gem` vs. `rubygem`

**Comment by ken-duck** [comment]
This is very curious. I suspect that at some time in the past we were moving from `gem` to `rubygem` as the type, and some code is stuck in a situation where it accepts either.

I would suggest that you use `rubygem` in all of your requests, and I will put in an internal issue report to remove the conversion from `gem` to `rubygem` on the website code to remove all confusion.

Thanks for pointing this out!

**Comment by csansone-handy** [comment]
Hi Ken - thanks for your reply! The [Package URL (PURL) Spec](https://github.com/package-url/purl-spec) doesn't exactly state that using `gem` is mandatory as the `type` for Ruby gems but it would be nice if that is what was used in OSS Index since that is what all the examples in the PURL spec use. Furthermore, tools such as [CycloneDX](https://cyclonedx.org) emit their Bill of materials using PURLs with `pkg:gem/*` and work very nicely with [OWASP Dependency Track](https://dependencytrack.org/) which uses OSS Index as (one of) its sources of vulnerability information.

To summarize: if you could settle on `gem` as the `type` for Ruby gems in your PURLs that would be awesome!

**Comment by ken-duck** [comment]
Ohhh. I see. Yes, that would seem to be the appropriate way to go.

**Comment by stevespringett** [comment]
FYI, the [Package URL test suite](https://github.com/package-url/purl-spec/blob/master/test-suite-data.json) used by all official Package URL implementations, use `gem`.

All implementations will use `gem`, not `rubygem`. I went back in history to Nov 11, 2017 when the original draft spec was published and it also stated `gem`. There doesn't appear to be anything in the history that would indicate that it was ever `rubygem`, so I would classify this issue as a defect.

**Comment by ken-duck** [comment]
That's hard to argue against :)

The original OSS Index internal representation said "rubygem" and predated our support for PURL. That's my excuse and I'm sticking to it ;)

**Comment by csansone-handy** [comment]
Just wanted to check if there was a decision made on this. @ken-duck did you decide if this is indeed an issue and you will be changing everything to reference `gem`?

**Comment by ken-duck** [comment]
Since it seems that the official type is `gem` we will be moving that way. I cannot yet speak to when it will be scheduled, though, but I am hoping to at least get it on the list sometime this week.

**Comment by csansone-handy** [comment]
Awesome thanks for clarifying! Looking forward to it!

**Comment by csansone-handy** [comment]
Just wanted to check if this has been prioritized

**Comment by ken-duck** [comment]
The PR is just waiting for review. Fingers crossed :)

**Comment by ken-duck** [comment]
I have tested this on our production server, it seems to be working. Note that both `gem` and `rubygem` work, as we would likely break some folks code by outright changing it.

**Comment by csansone-handy** [comment]
Looks good from an API perspective which is what matters most for my use case (OWASP Dependency Track pulling from the OSS Index API using `gem`). However, just note that only the `rubygem` listing appears on the vulnerability specific page in the UI. Note sure if this is intended or a separate issue

**Comment by ken-duck** [comment]
Bah. Forgot that bit. I'll change that up too.

**Comment by ken-duck** [comment]
We are going through some old issues. I have been informed that this issue has been resolved, probably for some time now.
